[WebSockets] Implement updates to use websockets onchain on Shannon protocol and remove websockets URL bypass

commoddity opened the pull request:
## Summary

Implement Shannon protocol endpoint selection for WebSocket nodes. Remove temporary WebSocket bypass.

https://github.com/user-attachments/assets/12ddfa88-3857-4084-a4f3-69ccc132ac24

### Primary Changes:
- Update `gateway` request handling to use protocol endpoint selection process.
- Removed `WebsocketEndpoints` configuration field from the Shannon protocol config to remove temporary hardcode websockets URL.
- Ensure RelayMiner connection closing propagates to Client to allow Client to handle reconnection on their end.

### Secondary changes:
- Refactored websocket bridge for code health.
- Removed stop channel in bridge in favour of using context cancellation.
- Updated all websockets unit tests.

## Not Included (NEXT UP)
- Update QoS to ensure selected endpoint supports websocket requests before selecting.
- Update documentation to include WebSockets.

**Please review the changes and provide feedback as needed.**

## Type of change

Select one or more from the following:

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Sanity Checklist

- [ ] I have updated the GitHub Issue 'assignees', 'reviewers', 'labels', 'project', 'iteration' and 'milestone'
- [ ] For docs, I have run 'make docusaurus_start'
- [ ] For code, I have run 'make test_all'
- [ ] For configurations, I have update the documentation
- [ ] I added TODOs where applicable

Olshansk reviewed (changes requested):
Just a few NITs, but I don't have enough context (yet) to provide strong feedback.

Key edits:
1. Please resole conflicts with main
2. I couldn't fully understand if an EVM websocket would leverage the `qos/evm` package for free out of the box?

Olshansk reviewed (changes requested):
Just a few NITs, but I don't have enough context (yet) to provide strong feedback.

Key edits:
1. Please resole conflicts with main
2. I couldn't fully understand if an EVM websocket would leverage the `qos/evm` package for free out of the box?

commoddity commented:
> Just a few NITs, but I don't have enough context (yet) to provide strong feedback.
> 
> Key edits:
> 
> 1. Please resole conflicts with main
> 2. I couldn't fully understand if an EVM websocket would leverage the `qos/evm` package for free out of the box?

After Arash's target branch got merged to `main` there were a ton of merge conflicts and way more changed lines; I've fixed the conflicts and updated from `main` and now changes lines were around ~1200 to ~500.

commoddity commented on local/grafana-dashboards/service-requests.json:
```diff
@@ -484,6 +484,103 @@
         "x": 0,
         "y": 18
       },
+      "id": 2,
```
Nothing; this was appearing as a change due to conflicts with Arash's branch. It's outdated now.

commoddity commented on qos/evm/response_generic.go:
```diff
@@ -44,7 +44,7 @@ type responseGeneric struct {
 	// Why the response has failed validation.
 	// Used when generating observations.
-	invalidReason *qosobservations.EVMResponseInvalidReason
+	validationErrorKind *qosobservations.EVMResponseValidationErrorKind
```
Ditto above; this is from the other branch and is now outdated.

commoddity commented on websockets/connection.go:
```diff
@@ -52,46 +60,68 @@ type message struct {
 	}
 
-	conn, _, err := websocket.DefaultDialer.Dial(u.String(), http.Header{})
+	// TODO_TECHDEBT(@commoddity): Remove this switch once RelayMiner
```
Relay Miner only allows endpoints with `http|https` protocols so we have to perform some hackery to switch the protocols in the code right now. I believe @red-0ne is aware of this as well so perhaps we could update that before we merge both this and the Relay Miner PR?

Olshansk commented on websockets/connection.go:
```diff
@@ -52,46 +56,59 @@ type message struct {
-		source:   source,
-		msgChan:  msgChan,
-		stopChan: stopChan,
+		ctx:       ctx,
```
NIT: Please group / space this in the same way as the struct

Olshansk reviewed (approved):
LGTM but please wait for a review + approval from @red-0ne before merging in!

red-0ne reviewed (approved):
Great refactoring for Shannon WebSockets support! :electric_plug: 

Approving the merge of this PR, but could you please add a TODO based on the comment I left?

We can sync on this next week
